compiler/src/dotty/tools/dotc/core/Phases.scala

```diff
@@ -333,16 +336,25 @@ object Phases {
     def subPhases: List[Run.SubPhase] = Nil
     final def traversals: Int = if subPhases.isEmpty then 1 else subPhases.length
 
+    /** skip the phase for a Java compilation unit, may depend on -Ypickle-java */
+    def skipIfJava(using Context): Boolean = true
+
     /** @pre `isRunnable` returns true */
     def run(using Context): Unit
 
     /** @pre `isRunnable` returns true */
     def runOn(units: List[CompilationUnit])(using runCtx: Context): List[CompilationUnit] =
       val buf = List.newBuilder[CompilationUnit]
+      val skipJavaUnits =
+        ctx.settings.YpickleJava.value && ctx.phase <= sbtExtractApiPhase && skipIfJava
```

Comment:
I don't understand why this is not just `skipIfJava`.

Reply:
it avoids constantly checking `typedAsJava` on every unit when not needed (e.g. its usually true, so usually we would unnecessarily check if the unit is java), otherwise this is only computed once

Reply:
there's a comment explaining this in https://github.com/lampepfl/dotty/pull/19074